js/everything.js

```diff
@@ -428,24 +489,104 @@ function anchorURLs(text) {
         // We have already determined the video id
-        return currentVideoID;
-    }
-
-    var info = parseYoutubeVideoID(url);
-
-    if (info.id) {
-        currentVideoID = info.id;
-        ga("send", "event", "video ID format", info.format);
-        return info.id;
+        callback.success(currentVideoID);
     }
     else {
-        errorMessage.show("Failed to parse the video ID.");
+        // First try parse Soundcloud id from URL.
+        parseSoundcloudVideoID(
+            url,
+            soundcloudClientID,
+            {
+                success: function(info) {
+                    // Success parsed ID from Soundcloud URL.
+                    ga("send", "event", "video ID format", info.format);
+                    currentVideoID = info.id;
+                    callback.success(info.id);
+                },
+                fail: function(info) {
+                    // Failed, so next try parse Youtube ID from URL.
+                    info = parseYoutubeVideoID(url);
+                    if (info.id) {
+                        // Success parsed Youtube ID.
+                        currentVideoID = info.id;
+                        callback.success(info.id);
+                    }
+                    else {
+                        // No Soundcloud or Youtube ID.
+                        callback.fail();
+         //               errorMessage.show("Failed to parse the video ID.");
+                    }
+                }
+            });
     }
 }
 
+function verifyYoutubeID(id, callback) {
+    $.ajax({
+        url: "https://www.googleapis.com/youtube/v3/videos",
+        dataType: "json",
+        data: {
+            key: youTubeDataApiKey,
+            part: "snippet",
+            fields: "items/snippet/description",
+            id: id
+        },
+        success: function(data) {
+            if (data.items.length === 0) {
+                callback.fail(false);
+            }
+            else {
+                callback.success(true);
+            }
+        }
+    }).fail(function() {
+        callback.fail(false);
+    });
+}
+
+function verifySoundcloudID(id, callback) {
+    $.ajax({
+        url: "https://api.soundcloud.com/tracks/" + id + "?client_id=" + soundcloudClientID,
+        dataType: "json",
+        success: function(data) {
+            if (data.id.toString() === id.toString()) {
+                callback.success(true);
+            }
+            else {
+                callback.fail(false);
+            }
+        }
+    }).fail(function() {
+        callback.fail(false);
+    });
+}
+
+// Verifies ID for youtube and soundcloud
+function verifyID(id, callback) {
+    verifyYoutubeID(
+        id,
+        {
+            success: function() {
+                callback.success();
+            },
+            fail: function() {
+                verifySoundcloudID(
+                    id,
+                    {
+                        success: function() {
```

Comment:
If we're the only consumers of these `success` and `fail` functions, we might just pass another parameter instead of passing objects around